### PR TITLE
Fix max table width calculation

### LINKED ISSUE #288: max_table_width does not work as expected
### What did you do?
Set max_table_width on a table to `os.get_terminal_size().columns` and played around with terminal size.
 
### What did you expect to happen?
The table should not exceed max_table_width (no line breaks in terminal)

max_width_should be max. 

### What actually happened?
Sometimes tables are wider than terminal size and i see line breaks

### What versions are you using?

* OS: MacOs 
* Python: 3.12
* PrettyTable: 3.10.0

### PR DESCRIPTION
**Pull request by stuertz**
Try to fix max_table_width computation

fixes: #288

### CODE REVIEW

**Comment by stuertz** [comment]
Hi @hugovk  is something missing here? Are you waiting for something to be added?

**Comment by hugovk** [comment]
Hi! Thank you very much for the PR! I'm just waiting to find time to review it :) I'll definitely try to get to it this month.

File: src/prettytable/prettytable.py [hunk 1]
```diff
@@ -1590,21 +1595,23 @@ def _compute_widths(self, rows, options) -> None:
...
             if table_width > self._max_table_width:
                 # Shrink widths in proportion
-                scale = 1.0 * self._max_table_width / table_width
-                widths = [int(w * scale) for w in widths]
+                vrulers = len(widths) - 1
+                scale = (
+                    self._max_table_width - per_col_padding * len(widths) - vrulers
+                ) / (table_width - per_col_padding * len(widths) - vrulers)
+                widths = [max(1, int(w * scale)) for w in widths]
                 self._widths = widths
```
Comment (hugovk) [inline src/prettytable/prettytable.py]: We can combine these:
```suggestion
                self._widths = [max(1, int(w * scale)) for w in widths]
```

File: src/prettytable/prettytable.py [hunk 2]
```diff
@@ -1590,21 +1595,23 @@ def _compute_widths(self, rows, options) -> None:
...
 
+        per_col_padding = sum(self._get_padding_widths(options))
         # Are we exceeding max_table_width?
         if self._max_table_width:
             table_width = self._compute_table_width(options)
             if table_width > self._max_table_width:
                 # Shrink widths in proportion
-                scale = 1.0 * self._max_table_width / table_width
-                widths = [int(w * scale) for w in widths]
+                vrulers = len(widths) - 1
```
Comment (hugovk) [inline src/prettytable/prettytable.py]: Can you think of another name for `vrulers`? It's not clear from the name what it's for.

File: src/prettytable/prettytable.py [hunk 3]
```diff
@@ -1590,21 +1595,23 @@ def _compute_widths(self, rows, options) -> None:
...
+                scale = (
+                    self._max_table_width - per_col_padding * len(widths) - vrulers
+                ) / (table_width - per_col_padding * len(widths) - vrulers)
```
Comment (hugovk) [inline src/prettytable/prettytable.py]: Both numerator and denominator use `per_col_padding * len(widths) - vrulers`.

Shall we move this out to a variable pre-calculating `per_col_padding * len(widths) + vrulers`?

Won't really affect performance, but will make the `scale` line easier to read.

We could also pre-calculate as `per_col_padding * len(widths) + len(widths) - 1` and remove `vrulers`.

**Review by hugovk — COMMENTED** [review]
Thank you for this! This is mainly style/readability suggestions.

File: src/prettytable/prettytable.py (hunk 2 above)
Comment (stuertz) [reply src/prettytable/prettytable.py]: It's the number of vertical chars or how do you call `|`. What about `no_of_vchars`

Comment (hugovk) [reply src/prettytable/prettytable.py]: Sure, `no_of_vchars` or `num_vchars` or `n_vchars` or something?

File: src/prettytable/prettytable.py (hunk 3 above)
Comment (stuertz) [reply src/prettytable/prettytable.py]: yep. I'll chose `markup_chars` as varname for the param

**Comment by hugovk** [comment]
Thank you very much!

**Pull request merged**